### ISSUE TITLE
Overlapping NavigationRegion3Ds make NavigationLink3Ds unusable

### ISSUE THREAD

**Issue by venilark** [godotengine/godot#80172]
### Godot version

4.1.1

### System information

Windows 10

### Issue description

I want 2 different type of NavigationAgent3Ds to use two different NavigationRegions, which also use different NavigationLink3Ds, all layers are, or should be correctly set. Problem is the first NavigationRegion3D will be detected when using the NavigationLinks. 

### Steps to reproduce

Press play, see how AgentBig can reach the destination but AgentSmall doesn't move. Now put NavigationRegion3DSmall BEFORE NavigationRegion3DBig to see the oposite effect. It seems only the NavigationLinks with the layer of the first NavigationRegion3D are detected.

### Minimal reproduction project

[navlinks_jump_test.zip](https://github.com/godotengine/godot/files/12241713/navlinks_jump_test.zip)

**Comment by venilark** [comment]
If you place the Destination node close to the ground, then both agents can reach it, problem seems to happen only when links are involved.

**Comment by smix8** [comment]
You can not stack multiple overlapping navigation meshes, or links for that matter, on the same navigation map in the same place. That will create a giant navigation mesh mess were nothing works anymore. 


Only two navigation mesh edges can ever merge together with the same edgekey when rasterized by the navigation map.
A navigation link position can only connect to a single polygon, not multiple.

See documentation for connection navigation meshes correctly [here]( https://docs.godotengine.org/en/stable/tutorials/navigation/navigation_connecting_navmesh.html) (no overlap) and documentation what the navigation layers are for and do [here]( https://docs.godotengine.org/en/stable/tutorials/navigation/navigation_using_navigationlayers.html) and documentation for multiple agent sizes [here](https://docs.godotengine.org/en/stable/tutorials/navigation/navigation_different_actor_types.html).

**Comment by venilark** [comment]
> 

So if I got it right from the last link, I can create the NavigationMesh from code or I could create the different NavigationMeshes with the editor, bake them to see how they look like, save them in a folder and finally use those resources like in the example code, so each agent can call the appropiate NavigationMesh using the corresponding Map. Would the NavigationLinks be a problem if they are overlapping or something?

**Comment by smix8** [comment]
You can overlap all you want in the Editor as long as the navigation related objects are later placed on different navigation maps. 

E.g. in your example a script could move the NavigationRegion3DSmall to a different navigation map at scene start and the custom NavigationAgent script for the small agent would set the navigation map of the agent to this small navigation map.  Same for the links that should be used by the small agent pathfinding.

All navigation related objects have functions to change the navigation map in script or if not, you can get their RID and change it with the NavigationServer API.

Now for pathfinding this works fine with different navigation maps, just query a path from the map that you need or switch NavigationAgent to the map it should use.

For avoidance to get working when each agent size has its own pathfinding map you would need to create a new shared avoidance map that has all agents place on it so they can "see" and avoid each other. Basically each actor would have two internal NavigationServer agents, the one used for the specific pathfinding map and the one used for the shared avoidance map.

**Comment by venilark** [comment]
> You can overlap all you want in the Editor as long as the navigation related objects are later placed on different navigation maps.
> 
> You can think of maps as with real world maps, one is for pedestrians, one for cars, one for trucks. They are all in the same world with the same geometry but crafted for specific "sizes" and you look / query paths from the map for the size you / the agent have.
> 
> E.g. in your example a script could move the NavigationRegion3DSmall to a different navigation map at scene start and the custom NavigationAgent script for the small agent would set the navigation map of the agent to this small navigation map. Same for the links that should be used by the small agent pathfinding.
> 
> All navigation related objects have functions to change the navigation map in script or if not, you can get their RID and change it with the NavigationServer API.
> 
> Now for pathfinding this works fine with different navigation maps, just query a path from the map that you need or switch NavigationAgent to the map it should use.
> 
> For avoidance to get working when each agent size has its own pathfinding map you would need to create a new shared avoidance map that has all agents place on it so they can "see" and avoid each other. Basically each actor would have two internal NavigationServer agents, the one used for the specific pathfinding map and the one used for the shared avoidance map.

I'm having some trouble understanding the avoidance part.
Right now I need to create the 2 different maps using the last link you provided (CSG problems aside https://github.com/godotengine/godot/issues/79843). Now to make avoidance work what should be done exactly? Create a third map common for both and change the map from small/big to "common" before calling navigation_agent.set_velocity()?

-Small calls set_navigation_map(small_map)
-Small calls is_target_reachable()/get_next_path_position()/whatever with the current map
-finally Small calls set_navigation_map(common_map) and then navigation_agent.set_velocity()?
-repeat

Is that it? In that case, what should the cell size, radius etc of the Agent of the common map be, the one from Small or Big?

Also unsure if I need to add a second NavigationAgent3D node after what I read, this whole process is a bit confusing.

**Comment by smix8** [comment]
Avoidance is calculated per map for all active avoidance agents on the map. So if the agents use different navigation maps they will need to share at least one map for the avoidance to still work.

It is not impossible to do this and still use NavigationAgent nodes but it is imo more difficult to use the nodes compared to just creating a custom agent that does everything from scratch in a script. The `NavigationAgent` nodes can not be split between maps by default as they have only 1 NavigationServer agent RID internally.

Since an agent can only ever join a single navigation map. as if an agent joins another navigation map it leaves the old navigation map, you will need to create a new agent in script and update it with scripts in parallel.

Use `NavigationServer3D.agent_create()` to create the new agent RID dedicated to avoidance.
Place this new agent on the new "avoidance" map with `NavigationServer3D.agent_set_map()`.

Now update the agent position and agent velocity every `physics_process` frame with `NavigationServer3D.agent_set_position()` and `NavigationServer3D.agent_set_velocity()`. Also create a custom avoidance callback for the agent with `NavigationServer3D.agent_set_avoidance_callback()` so you can receive the safe_velocity somewhere.

The agent radius for the avoidance should be the size what your agent is as avoidance simulation recognized the size of avoidance objects, only the pathfinding and navmesh does not.

**Comment by smix8** [comment]
Looks reasonable but I did not test it.

Note that `navigation_agent.is_target_reachable()` does a full internal pathfinding to determine if the target is reachable so it will cost you a lot of performance doing this every single physics frame.

You should only query this a single time per navigation map iteration or not query at all.

If you just set the target_position you can look at the returned path and if the last path position is too far away from your original target you will know that the target is not reachable completely. This way you do not need to perform 2 path searches back to back.

**Comment by venilark** [comment]
> Looks reasonable but I did not test it.
> 
> Note that `navigation_agent.is_target_reachable()` does a full internal pathfinding to determine if the target is reachable so it will cost you a lot of performance doing this every single physics frame.
> 
> You should only query this a single time per navigation map iteration or not query at all.
> 
> If you just set the `target_position` you can look at the returned path and if the last path position is too far away from your original target you will know that the target is not reachable completely. This way you do not need to perform 2 path searches back to back.

good idea, I'll keep that in mind
with my code I can't parse the geometry in any mode (mesh/colliders) if there are CSG nodes so I'll wait until the other PR is merged to see if there is something wrong

**Comment by smix8** [comment]
@venilark You can parse CSG node when your NavigationMesh is set to parse visual meshes. The bug in the mentioned pr is only for collision parsing.

**Comment by venilark** [comment]
![c](https://github.com/godotengine/godot/assets/98270528/49b18ea1-992d-4867-9106-d93ac1921aff)
then I must have something else wrong because I already tried with both modes and the result is the same, always empty geometry, which doesn't happen when I try to bake in the editor with a NavigationRegion3D

**Comment by smix8** [comment]
CSG nodes work just fine, even with CSGCombiner3D involved. I just tested with master, no issue when the parsed geometry type is set to Mesh Instances. Must be an issue with your setup or the resulting mesh of the CSG operation.

**Comment by venilark** [comment]
[navlinks_jump_test.zip](https://github.com/godotengine/godot/files/12252469/navlinks_jump_test.zip)
copypasted the same code

**Comment by smix8** [comment]
The problem in that project is SceneTree init order of nodes, you need to use call_deferred because the majority of parsed nodes will not be ready while you run a function in a _ready() function.

**Comment by venilark** [comment]
> The problem in that project is SceneTree init order of nodes, you need to use `call_deferred` because the majority of parsed nodes will not be ready while you run a bake function in a `_ready()` function.
> 
> Even with deferred there are some nodes that take longer for their setup so they are not ready to be parsed on the first frame.
> 
> E.g. some physics nodes setup their collision shapes in init() so they actually have a shape at this point, but most nodes setup stuff in their own ready or when they join the SceneTree so their shapes and meshes will only be ready after the server did sync. Heavy nodes like TileMap or GridMap will do their entire cell setup deferred so you often need to wait a full frame until everything is ready to be parsed.

  ```
func _ready():
	  await get_tree().process_frame
	  _initialize_navigation_regions()
```
should this be enough for all usecases then?
the actors maybe have to delay one frame too the navigation but that's fine, as long as the map exists and doesn't throw any error

**Comment by smix8** [comment]
If I remember correctly this issue was later solved on other channels.

If there are no further questions I think we can close this issue.

**Comment by smix8** [comment]
Closing, feel free to reply if there are still open questions to this.
